src/lib/io/sol-memmap-storage-impl-zephyr.c

```diff
@@ -0,0 +1,285 @@
+        assert(entry->size <= 8);
+
+        for (i = 0, j = 0; i < entry->size; i++, j += 8)
+            value |= (uint64_t)((uint8_t *)blob->mem)[i] << j;
+
+        memcpy(&old_value, back_mem + mod, entry->size);
+
+        value <<= entry->bit_offset;
+        value &= mask;
+        value |= (old_value & ~mask);
+
+        memcpy(back_mem + mod, &value, entry->size);
+
+        r = flash_write_protection_set(flash_dev, false);
+        SOL_INT_CHECK_GOTO(r, < 0, error);
+
+        for (cnt = 0; cnt < back_sz; cnt += MAX_RW) {
+            r = flash_write(flash_dev, offset - mod + cnt, back_mem + cnt,
+                MAX_RW);
+            if (r < 0) {
+                SOL_WRN("Flash write failed");
+                goto error;
+            }
+            r = flash_write_protection_set(flash_dev, false);
+            SOL_INT_CHECK_GOTO(r, < 0, error);
+        }
+    } else {
+        size_t size;
+
+        if (blob->size > entry->size) {
+            SOL_WRN("Trying to store entry data of size %zu bytes to a entry "
+                "with %zu bytes of reserved space", blob->size, entry->size);
+        }
+        size = sol_util_min(entry->size, blob->size);
+
+        memcpy(back_mem + mod, blob->mem, size);
+
+        r = flash_write_protection_set(flash_dev, false);
+        SOL_INT_CHECK_GOTO(r, < 0, error);
+
+        for (cnt = 0; cnt < back_sz; cnt += MAX_RW) {
+            r = flash_write(flash_dev, offset - mod + cnt, back_mem + cnt,
+                MAX_RW);
+            if (r < 0) {
+                SOL_WRN("Flash write failed");
+                goto error;
+            }
+            r = flash_write_protection_set(flash_dev, false);
+            SOL_INT_CHECK_GOTO(r, < 0, error);
+        }
+    }
+
+    if (cb)
+        cb((void *)data, name, blob, r);
+
+    sol_blob_unref(blob);
+
+    return r;
+
+error:
+    SOL_DBG("Error writing to memmap: %s", sol_util_strerrora(-r));
+
+    if (cb)
+        cb((void *)data, name, blob, r);
+
+    sol_blob_unref(blob);
+
+    return r;
+}
+
+bool
+sol_memmap_impl_perform_pending_writes(void *data)
+{
+    struct map_internal *map_internal = data;
+    struct pending_write_data *pending;
+    struct sol_vector tmp_vector;
+    int i;
+
+    map_internal->timeout = NULL;
+
+    tmp_vector = map_internal->pending_writes;
+    sol_vector_init(&map_internal->pending_writes,
+        sizeof(struct pending_write_data));
+
+    SOL_VECTOR_FOREACH_IDX (&tmp_vector, pending, i) {
+        sol_memmap_impl_write_raw((struct map_internal *)map_internal,
+            pending->name, pending->entry, pending->mask, pending->blob,
+            pending->cb, pending->data);
+        free(pending->name);
+        sol_blob_unref(pending->blob);
+    }
+    sol_vector_clear(&tmp_vector);
+
+    SOL_DBG("Performed pending writes");
+
+    return false;
+}
+
+int
+sol_memmap_impl_init(void)
+{
+    if (flash_dev)
+        return 0;
+
+    flash_dev = device_get_binding("W25QXXDV");
```

Comment:
all in all I think the driver and parameters are either specified in the kconfig (not so good) or in the memmap itself. We can use the path as something like `<driver>,<erase_size>,<max_rw>,<offset>`, so in this case you'd use `W25QXXDV,4096,256,0`. You could make some of these optional, or even use key=value pairs, like `driver=W25QXXV,erase_size=4096,max_rw=256`. Then you keep these in the context of memmap, not as a global, pass the handle and get the information from there.

I'd talk to the Zephyr guys if these informations (erase, max_rw) could be queried on their side, in this case avoid these options on our side altogether.

@ibriano how does that sound to you?